pwcnet/cost_volume/cost_volume.py

```diff
@@ -50,19 +46,91 @@ def cost_volume(c1, c2, search_range=4):
+    :param c1: Tensor. Feature map of shape [batch_size, H, W, num_features].
+    :param c2: Input tensor with the exact same shape as c1.
+    :param search_range: The search square's side length is equal to 2 * search_range + 1.
+    :return: Tensor. Cost volume of shape [batch_size, H, W, s * s], where s is equal to 2 * search_range + 1.
+    """
+    with tf.name_scope('cost_volume'):
+        square_len = 2 * search_range + 1
+        square_area = square_len ** 2
+        cv_shape = tf.concat([tf.shape(c1)[:-1], [square_area]], axis=0)
+
+        # The batch dimension needs to be moved to the end to make slicing work correctly.
+        c1_transposed = tf.transpose(c1, [1, 2, 3, 0])
+        c2_transposed = tf.transpose(c2, [1, 2, 3, 0])
+
+        # Form an index matrix to help us update sparsely later on.
+        cv_height, cv_width = cv_shape[1], cv_shape[2]
+        x_1d, y_1d, z_1d = tf.range(0, cv_width), tf.range(0, cv_height), tf.range(0, square_area)
+        x_2d, y_2d, z_2d = tf.meshgrid(x_1d, y_1d, z_1d)
+        indices_3d = tf.stack([y_2d, x_2d, z_2d], axis=-1)
+
+        all_indices, all_costs = [], []
+        all_c1_regions, all_c2_regions = [], []
+        cur_z_index = square_area - 1
+        num_channels = tf.shape(c1)[-1]
+        batch_size = tf.shape(c1)[0]
+        for i in range(-search_range, search_range + 1):
+            for j in range(-search_range, search_range + 1):
+
+                # Note that 'Python'[slice(None)] returns 'Python'.
+                if i < 0:
+                    slice_h, slice_h_r = slice(None, i), slice(-i, None)
+                elif i > 0:
+                    slice_h, slice_h_r = slice(i, None), slice(None, -i)
+                else:
+                    slice_h, slice_h_r = slice(None), slice(None)
+
+                if j < 0:
+                    slice_w, slice_w_r = slice(None, j), slice(-j, None)
+                elif j > 0:
+                    slice_w, slice_w_r = slice(j, None), slice(None, -j)
+                else:
+                    slice_w, slice_w_r = slice(None), slice(None)
+
+                # Get the coordinates for scatter update, where each element is a (y, x, z) coordinate.
+                cur_indices = indices_3d[slice_h, slice_w, cur_z_index]
+                cur_indices = tf.reshape(cur_indices, (-1, 3))
+                c1_region = c1_transposed[slice_h, slice_w, :, :]
+                c1_region = tf.reshape(c1_region, (-1, num_channels, batch_size))
+                c2_region = c2_transposed[slice_h_r, slice_w_r, :, :]
+                c2_region = tf.reshape(c2_region, (-1, num_channels, batch_size))
+                all_indices.append(cur_indices)
+                all_c1_regions.append(c1_region)
+                all_c2_regions.append(c2_region)
+                cur_z_index -= 1
+
+        all_indices = tf.concat(all_indices[::-1], axis=0)
+        all_c1_regions = tf.concat(all_c1_regions[::-1], axis=0)
```

Comment:
Is it not possible to use all_indices and a single gather_nd call to do all the slicing at once?

Reply:
I can try that and see if it helps.